src/osd/OSDMapRecovery.cc

```diff
@@ -0,0 +1,463 @@
+
+  while (!ranges.empty()) {
+    pair<epoch_t,epoch_t> r = ranges.front();
+    ranges.pop_front();
+
+    list<pair<epoch_t,epoch_t> > missing;
+    int ret = recover_range(r, missing, inc);
+    if (ret < 0) {
+      derr << __func__ << " error recovering range ["
+        << r.first << " .. " << r.second << "]" << dendl;
+      state = STATE_ERROR;
+      return ret;
+    }
+    if (!missing.empty()) {
+      ranges.splice(ranges.begin(), missing);
+    }
+  }
+  return 0;
+}
+
+int OSDMapRecovery::recover(const epoch_t first, const epoch_t last)
+{
+  assert(state == STATE_INIT);
+
+  state = STATE_RECOVERING;
+  dout(10) << __func__ << " epochs from " << first
+           << " to " << last << dendl;
+
+  list<pair<epoch_t,epoch_t> > broken_maps;
+  if (!find_broken_ranges(first, last, broken_maps, false)) {
+    dout(0) << __func__ << " no broken maps found" << dendl;
+  }
+
+  list<pair<epoch_t,epoch_t> > broken_inc_maps;
+  if (!find_broken_ranges(first, last, broken_inc_maps, true)) {
+    dout(0) << __func__ << " no broken inc maps found" << dendl;
+  }
+
+  int ret = _recover(broken_maps, false);
+  if (ret < 0) {
+    derr << __func__ << " error recovering broken maps" << dendl;
+    state = STATE_ERROR;
+    return ret;
+  }
+
+  ret = _recover(broken_inc_maps, true);
+  if (ret < 0) {
+    derr << __func__ << " error recovering broken inc maps" << dendl;
+    state = STATE_ERROR;
+    return ret;
+  }
+
+  state = STATE_DONE;
+  dout(0) << __func__ << " finished recovery" << dendl;
+  return 0;
+}
+
+bool OSDMapRecovery::_is_broken_map(epoch_t e)
+{
+  try {
+    bufferlist bl;
+
+    if (!osd->service.get_map_bl(e, bl, false) || bl.length() == 0) {
+      throw ceph::buffer::end_of_buffer();
+    }
+
+    OSDMap m;
+    m.decode(bl);
+
+  } catch (ceph::buffer::end_of_buffer e) {
+    return true;
+  } catch (ceph::buffer::malformed_input e) {
+    return true;
+  }
+  return false;
+}
+
+bool OSDMapRecovery::_is_broken_inc(epoch_t e)
+{
+  try {
+    bufferlist bl;
+
+    if (!osd->service.get_inc_map_bl(e, bl, false) || bl.length() == 0) {
+      throw ceph::buffer::end_of_buffer();
+    }
+
+    OSDMap::Incremental inc;
+    auto p = bl.begin();
+    inc.decode(p);
+
+  } catch (ceph::buffer::end_of_buffer e) {
+    return true;
+  } catch (ceph::buffer::malformed_input e) {
+    return true;
+  }
+
+  return false;
+}
+
+bool OSDMapRecovery::find_broken_ranges(
+    epoch_t first,
+    epoch_t last,
+    list<pair<epoch_t,epoch_t> >& ranges,
+    bool inc)
+{
+
+  dout(10) << __func__
+           << " " << (inc ? "inc" : "full" )
+           << " [" << first << " .. " << last << "]" << dendl;
+
+  for (auto e = first; e <= last; e ++) {
+
+    bool broken = (inc ? _is_broken_inc(e) : _is_broken_map(e));
+
+    if (broken) {
+      dout(0) << __func__ << " " << (inc ? "inc" : "full")
+              << " epoch " << e << " broken" << dendl;
+
+      auto &l = ranges.back();
```

Comment:
> Calling back on an empty container is undefined.

maybe we can note down the upper bound of missing/corrupted map? and only try to append to it when `ranges` is not empty or the upper bound is not "-1" or some sentry number?